torchao/quantization/prototype/qat/affine_fake_quantized_tensor.py

```diff
@@ -0,0 +1,253 @@
+            self.original_tensor.to(device),
+            self.apply_fake_quant_fn,
+            self.fake_quant_enabled,
+            **kwargs,
+        )
+
+    def _apply_fn_to_data(self, fn):
+        """
+        Create a new `AffineFakeQuantizedTensor` with `fn` applied to the
+        original tensor, to be called within __torch_dispatch__.
+        """
+        return self.__class__(
+            fn(self.original_tensor),
+            self.apply_fake_quant_fn,
+            self.fake_quant_enabled,
+            requires_grad=False,
+        )
+
+    implements = classmethod(_implements)
+    __torch_function__ = classmethod(_dispatch__torch_function__)
+    __torch_dispatch__ = classmethod(_dispatch__torch_dispatch__)
+
+implements = AffineFakeQuantizedTensor.implements
+
+
+@implements(torch.nn.functional.linear)
+def _(func, types, args, kwargs):
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+    if isinstance(input_tensor, AffineFakeQuantizedTensor):
+        input_tensor = input_tensor.get_value()
+    if isinstance(weight_tensor, AffineFakeQuantizedTensor):
+        weight_tensor = weight_tensor.get_value()
+    return torch.nn.functional.linear(input_tensor, weight_tensor, bias)
+
+
+@implements([aten.mm.default, aten.addmm.default])
+def _(func, types, args, kwargs):
+    if func == aten.addmm.default:
+        bias = args[0]
+        input_tensor = args[1]
+        weight_tensor = args[2]
+    else:
+        bias = None
+        input_tensor = args[0]
+        weight_tensor = args[1]
+    if isinstance(input_tensor, AffineFakeQuantizedTensor):
+        input_tensor = input_tensor.get_value()
+    if isinstance(weight_tensor, AffineFakeQuantizedTensor):
+        weight_tensor = weight_tensor.get_value()
+    if bias is not None:
+        return func(bias, input_tensor, weight_tensor)
+    else:
+        return func(input_tensor, weight_tensor)
+
+
+@implements([aten.detach.default])
+def _(func, types, args, kwargs):
+    return return_and_correct_aliasing(
+        func, args, kwargs, args[0]._apply_fn_to_data(torch.detach),
+    )
+
+
+@implements([aten.clone.default])
+def _(func, types, args, kwargs):
+    return return_and_correct_aliasing(
+        func, args, kwargs, args[0]._apply_fn_to_data(torch.clone),
+    )
+
+
+@implements([aten.t.default])
+def _(func, types, args, kwargs):
+    return return_and_correct_aliasing(
+        func, args, kwargs, args[0]._apply_fn_to_data(torch.t),
+    )
+
+
+@implements([
+    aten.add.Tensor,
+    aten.add_.Tensor,
+    aten.mul_.Tensor,
+])
+def _(func, types, args, kwargs):
+    assert len(args) == 2, f"dispatched the wrong op to the binary handler: {func}"
+    new_args = pytree.tree_map_only(AffineFakeQuantizedTensor, lambda x: x.original_tensor, args)
+    first_afq_tensor = args[0] if isinstance(args[0], AffineFakeQuantizedTensor) else args[1]
+    fn = lambda x: func(*new_args, **kwargs)
```

Comment:
it seems that `x` is ignored, looks weird

Reply:
I think the problem is for these ops, I found that sometimes args[1] is the subclass, so we need to handle both cases. There's a bit of complexity here so I feel it's better to do it in a common function

Reply:
I rewrote this a bit to use a new helper function that simply accepts a new value, hope that's clearer